### ISSUE TITLE
Suggestion: FlxInputText

### ISSUE THREAD

**Issue by SeiferTim** [FlixelCommunity/flixel#137]
a: I'm not sure where suggestions should go...
b: This is probably for a release after the bug-fix release

I've been using the FlxInputText Class for awhile and it really does pretty much as advertised and I haven't had any problems with it. Should think about including it in the core code...

I got my version from here: http://forums.flixel.org/index.php/topic,272.0.html and I think there have been changes to it since I started using it.

**Comment by moly** [comment]
This is a good idea, but the class should take this approach: http://forums.flixel.org/index.php/topic,272.msg35492.html#msg35492

The one Tim linked to draws the text by adding it directly to the stage, which causes incompatibility with some parts of flixel.

**Comment by Dovyski** [comment]
I like the idea too. 

@moly Could you tell some of the incompatibility the stage approach causes?

**Comment by moly** [comment]
The FlxText will always appear on top of everything else. I think that it also doesn't work with the cameras at all.

**Comment by Gama11** [comment]
Yeah, it's a great idea. I've missed textfields for flixel in the past and simply used a normal `TextField`, but it doesn't work too well in conjunction with flixel. The flixel cursor is under the textfield and I always have to remove it manually from `FlxG.stage` when switching states. 

According to wg/funstorm in that thread @moly posted, that class seems very incomplete though. It's probably the best approach there is right now though.

**Comment by Dovyski** [comment]
Ok, I got the problems. Thanks, @moly  and @Gama11 

I read the post @moly pointed out and also noticed the comments that say the class seems very incomplete. We can work on improving that class or create our own version.

**Comment by Gama11** [comment]
I've actually started working on improving the latest version of `FlxInputText` by Mr_Walrus in that thread. It actually seems quite solid, even though there are quite a few missing features and bug like wg/funstorm pointed out. I'll push it to a repo and create a little demo swf once I've fixed and added the most of the essential things. :)

**Comment by Gama11** [comment]
Made quite a bit of progress and pushed it to a repo of mine called [FlxInputText](https://github.com/Gama11/FlxInputText) with a [demo showing off the features of the class](https://dl.dropbox.com/u/18627942/FlxInputTextDemo.swf). (Not sure how to convince GitHub not to ignore the /bin folder :/)

I've made an inital commit with the original class by Mr_Walrus, so you can actually nicely see the changes I've made [here](https://github.com/Gama11/FlxInputText/commit/d94842d0d660bb057ebb5d898a176b4f88e75967).

Brief overview:

#### Changes:
- Turned `filterMode` into a getter / setter so the initial label that is set is also filtered
- Removed unused variables
- `setMaxLength` didn't actually work, changed it to getter and setter `maxLength`
- `caretTimer` doesn't start twice anymore and blink extremely fast
- Added a `FlxSprite` representing the borders
- Added `borderColor`, `borderThickness` getters / setters
- Made the height of the `caret` variable depending on the text `size`
- Overriden `size` setter to change `caret size`
- Textfield's height won't increase anymore by default when `text length` exceeds space
- `forceUpperCase` is now `forceCase` with static variables `ALL_CASES`, `UPPER_CASE` and `LOWER_CASE`
- Added `passwordMode` getter and setter for password textfields
- Made sure the `caret` doesn't exceed the textfields `width` if single-lined
- Disabled shift, ctrl and key 220 (opens flixel console) for textfield input. Shift and ctrl would cause weird results

#### What's not there yet:
- Multi-line textfields are buggy. Couldn't figure out how to correctly calculate the `height` of multi-line textfields for the background, though there's a setter `lines` which is supossed to lock the textfield's size to a specific height
- Text can't be selected with the mouse
- You can write text in single-line textfields even though it isn't shown anymore

I'm open to any criticism! :) Probably forgot mentioning some changes.

**Comment by Dovyski** [comment]
That's amazing, @Gama11 ! :) Thank you!

Your version of `FlxInputText` looks pretty solid. I think a few adjustments and it would be a great addition to Flixel. I noticed another feature that is missing:
- Moving the caret does not allow you to navigate in the text. If the text is too big, there is no way to read the words outside the textfield area.

**Comment by Gama11** [comment]
@Dovyski Glad you like it! :)

> - Moving the caret does not allow you to navigate in the text. If the text is too big, there is no way to read the words outside the textfield area.

is basically what I meant with:

> - You can write text in single-line textfields even though it isn't shown anymore

I have to admit, I probably didn't express it in a very clear way. Either way, I'm not a hundred percent sure which textfield behaviour is preferable, there are two possibilities:
1. Text scrolls with the caret's index
2. Max amount of characters a textfield can hold is limited to its size - all text will always be shown. Could be done via `maxLength`

I'm thinking 1. is probably the better way to go about it. What do you think? Also, do we need to ability to select text with the mouse, as well as Ctrl + A, X, C, V - support?

**Comment by Dovyski** [comment]
> You can write text in single-line textfields even though it isn't shown anymore

Oh, now I got it. Sorry for that!

Both solutions you provided for single-line textfields are good to me, but option 2 is very limited. I would go with option 1 (which is not that hard to implement).

About the ability to select text, I imagine it is a very tricky thing to implement. It would be great to have all those mouse/Ctrl+A/Ctrl+C/etc features, but if we consider Flixel a "retro" engine, there was no text selection back then :) I think that's not a mandatory feature.
